File: DraftRightWindows/DraftRightWindows/Services/ErrorReporter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace DraftRightWindows.Services;

/// <summary>
/// Sends unhandled exceptions to the DraftRight backend's /errors
/// endpoint. Pairs with App.cs's existing local crash logging — every
/// crash gets BOTH a local file and a server report.
///
/// Privacy: never sends user-typed text. Only error type, stack, and
/// a small context (OS version, app version, locale).
/// </summary>
public static class ErrorReporter
{
    private static readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(10) };
    private static string _backendUrl = "https://api.draftright.info";
    private static Func<string?>? _bearerTokenProvider;
    private static readonly string _queuePath = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "DraftRight", "error_queue.json");

    public static void Configure(string backendUrl, Func<string?>? bearerTokenProvider)
    {
        _backendUrl = backendUrl.TrimEnd('/');
        _bearerTokenProvider = bearerTokenProvider;

        // Flush anything stranded from a previous run
        _ = Task.Run(FlushPersistedQueue);
    }

    /// <summary>
    /// Fire-and-forget report. Falls back to a local queue if the network
    /// is down; the queue is drained on next startup.
    /// </summary>
    public static void Report(Exception ex, string source = "unknown", string severity = "error")
    {
        try
        {
            var report = Build(ex, source, severity);
            _ = SendOrQueue(report);
        }
        catch
        {
            // Reporting must never throw on the calling thread. Swallow.
        }
    }

    public static void ReportHandled(Exception ex, string severity = "warning")
    {
        Report(ex, source: "handled", severity: severity);
    }

    private static Dictionary<string, object?> Build(Exception ex, string source, string severity)
    {
        var asm = System.Reflection.Assembly.GetExecutingAssembly();
        var version = asm.GetName().Version?.ToString() ?? "?";

        return new Dictionary<string, object?>
        {
            ["platform"] = "windows",
            ["app_version"] = version,
            ["severity"] = severity,
            ["error_type"] = Truncate(ex.GetType().FullName ?? ex.GetType().Name, 200),
            ["message"] = Truncate(ex.Message ?? "", 5000),
            ["stack_trace"] = Truncate(ex.StackTrace ?? "", 20000),
            ["context"] = new Dictionary<string, object?>
            {
                ["source"] = source,
                ["os"] = Environment.OSVersion.VersionString,
                ["arch"] = System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture.ToString(),
                ["dotnet"] = Environment.Version.ToString(),
                ["locale"] = System.Globalization.CultureInfo.CurrentCulture.Name,
                ["ts"] = DateTime.UtcNow.ToString("o"),
            },
        };
    }

    private static string Truncate(string s, int max) =>
        s.Length > max ? s.Substring(0, max) : s;

    private static async Task SendOrQueue(Dictionary<string, object?> report)
    {
        var json = JsonSerializer.Serialize(report);
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, $"{_backendUrl}/errors");
            req.Content = new StringContent(json, Encoding.UTF8, "application/json");

            var token = _bearerTokenProvider?.Invoke();
            if (!string.IsNullOrEmpty(token))
                req.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);

            var resp = await _http.SendAsync(req).ConfigureAwait(false);
            if (resp.IsSuccessStatusCode) return;
            // Server rejected → drop. Don't queue malformed reports forever.
        }
        catch
        {
            // Network failed — persist for next launch
            PersistToQueue(json);
        }
    }

    private static void PersistToQueue(string json)
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_queuePath)!);
            var existing = File.Exists(_queuePath)
                ? File.ReadAllLines(_queuePath)
                : Array.Empty<string>();
            var trimmed = existing.Length > 100 ? existing[^100..] : existing;
            File.WriteAllLines(_queuePath, [..trimmed, json]);
        }
        catch
        {
            // Persistence is best-effort
        }
    }

    private static async Task FlushPersistedQueue()
    {
        if (!File.Exists(_queuePath)) return;
        try
        {
            var lines = File.ReadAllLines(_queuePath);
            File.Delete(_queuePath); // optimistic — we'll re-queue failures
            var remaining = new List<string>();

            foreach (var line in lines)
            {
                try
                {
                    using var req = new HttpRequestMessage(HttpMethod.Post, $"{_backendUrl}/errors");
                    req.Content = new StringContent(line, Encoding.UTF8, "application/json");

                    var token = _bearerTokenProvider?.Invoke();
                    if (!string.IsNullOrEmpty(token))
                        req.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);

                    var resp = await _http.SendAsync(req).ConfigureAwait(false);
                    if (!resp.IsSuccessStatusCode) remaining.Add(line);
                }
                catch
                {
                    remaining.Add(line);
                }
            }

            if (remaining.Count > 0)
            {
                Directory.CreateDirectory(Path.GetDirectoryName(_queuePath)!);
                File.WriteAllLines(_queuePath, remaining);
            }
        }
        catch
        {
            // ignore
        }
    }
}

File: DraftRightLinux/draftright/services/error_reporter.py
"""Server-side error reporting for the Linux GTK app.

Hooks ``sys.excepthook`` and threading/asyncio handlers, and POSTs
unhandled exceptions to the DraftRight backend's ``/errors`` endpoint.
Pairs with the existing local logger — every crash gets BOTH a local
log line and a server-side row.

Privacy: never sends user-typed text. Only error type, stack trace,
and a small context (Python version, GTK version, locale).
"""
from __future__ import annotations

import json
import logging
import platform
import os
import sys
import threading
import traceback
from datetime import datetime, timezone
from pathlib import Path
from typing import Callable, Optional
from urllib import request as urllib_request
from urllib.error import URLError

logger = logging.getLogger(__name__)

_BACKEND_URL: str = "https://api.draftright.info"
_BEARER_TOKEN_PROVIDER: Optional[Callable[[], Optional[str]]] = None
_QUEUE_PATH = Path(os.path.expanduser("~/.config/draftright/error_queue.jsonl"))
_APP_VERSION = "linux"


def configure(
    backend_url: str,
    bearer_token_provider: Optional[Callable[[], Optional[str]]] = None,
    app_version: str = "linux",
) -> None:
    """Wire up handlers + flush any queue stranded from a prior run."""
    global _BACKEND_URL, _BEARER_TOKEN_PROVIDER, _APP_VERSION
    _BACKEND_URL = backend_url.rstrip("/")
    _BEARER_TOKEN_PROVIDER = bearer_token_provider
    _APP_VERSION = app_version

    # sys.excepthook for top-level uncaught exceptions
    sys.excepthook = _excepthook

    # threading.excepthook for crashes on Thread targets (Python 3.8+)
    if hasattr(threading, "excepthook"):
        threading.excepthook = _thread_excepthook

    # asyncio doesn't get installed unless there's a running loop, but
    # we set the default exception handler factory; consumers using
    # asyncio.run() will pick this up.
    try:
        import asyncio
        asyncio.events.set_event_loop_policy(_AsyncioPolicy())
    except Exception:
        pass

    # Flush any persisted reports
    threading.Thread(target=_flush_queue_async, daemon=True).start()


def report(exc: BaseException, *, source: str = "unknown",
           severity: str = "error") -> None:
    """Manually report a caught-but-noteworthy exception."""
    try:
        payload = _build(exc, source=source, severity=severity)
        _send_or_queue(payload)
    except Exception:
        # Reporting must never throw on the calling thread
        pass


def report_handled(exc: BaseException, *, severity: str = "warning") -> None:
    report(exc, source="handled", severity=severity)


# ── Internals ────────────────────────────────────────────────────────────

def _excepthook(exc_type, exc, tb):
    try:
        report(exc, source="sys.excepthook", severity="fatal")
    except Exception:
        pass
    # Preserve the default printing behavior so the user still sees the trace
    sys.__excepthook__(exc_type, exc, tb)


def _thread_excepthook(args):
    try:
        if args.exc_value is not None:
            report(args.exc_value, source=f"thread:{args.thread.name}", severity="error")
    except Exception:
        pass


class _AsyncioPolicy:
    """Default asyncio policy that installs a custom exception handler."""
    def get_event_loop(self):
        import asyncio
        loop = asyncio.new_event_loop()
        loop.set_exception_handler(self._handle_asyncio)
        return loop

    @staticmethod
    def _handle_asyncio(loop, ctx):
        exc = ctx.get("exception")
        if exc:
            report(exc, source="asyncio", severity="error")


def _build(exc: BaseException, *, source: str, severity: str) -> dict:
    tb_lines = traceback.format_exception(type(exc), exc, exc.__traceback__)
    stack = "".join(tb_lines)[-20000:]  # tail-truncate
    return {
        "platform": "linux",
        "app_version": _APP_VERSION,
        "severity": severity,
        "error_type": type(exc).__name__[:200],
        "message": str(exc)[:5000],
        "stack_trace": stack,
        "context": {
            "source": source,
            "python": platform.python_version(),
            "system": platform.system(),
            "release": platform.release(),
            "machine": platform.machine(),
            "locale": os.environ.get("LANG", ""),
            "ts": datetime.now(timezone.utc).isoformat(),
        },
    }


def _send_or_queue(payload: dict) -> None:
    body = json.dumps(payload).encode("utf-8")
    try:
        req = urllib_request.Request(
            f"{_BACKEND_URL}/errors",
            data=body,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        token = _BEARER_TOKEN_PROVIDER() if _BEARER_TOKEN_PROVIDER else None
        if token:
            req.add_header("Authorization", f"Bearer {token}")
        with urllib_request.urlopen(req, timeout=10) as resp:
            if 200 <= resp.status < 300:
                return
    except (URLError, OSError, TimeoutError):
        _persist_to_queue(body)


def _persist_to_queue(body: bytes) -> None:
    try:
        _QUEUE_PATH.parent.mkdir(parents=True, exist_ok=True)
        existing = (
            _QUEUE_PATH.read_text().splitlines() if _QUEUE_PATH.exists() else []
        )
        # Trim oldest entries beyond 100
        if len(existing) > 100:
            existing = existing[-100:]
        existing.append(body.decode("utf-8"))
        _QUEUE_PATH.write_text("\n".join(existing) + "\n")
    except Exception:
        pass


def _flush_queue_async() -> None:
    if not _QUEUE_PATH.exists():
        return
    try:
        lines = _QUEUE_PATH.read_text().splitlines()
        _QUEUE_PATH.unlink(missing_ok=True)  # optimistic
        remaining = []
        for line in lines:
            line = line.strip()
            if not line:
                continue
            try:
                req = urllib_request.Request(
                    f"{_BACKEND_URL}/errors",
                    data=line.encode("utf-8"),
                    headers={"Content-Type": "application/json"},
                    method="POST",
                )
                token = _BEARER_TOKEN_PROVIDER() if _BEARER_TOKEN_PROVIDER else None
                if token:
                    req.add_header("Authorization", f"Bearer {token}")
                with urllib_request.urlopen(req, timeout=10) as resp:
                    if not (200 <= resp.status < 300):
                        remaining.append(line)
            except Exception:
                remaining.append(line)

        if remaining:
            _QUEUE_PATH.parent.mkdir(parents=True, exist_ok=True)
            _QUEUE_PATH.write_text("\n".join(remaining) + "\n")
    except Exception:
        pass
